### ISSUE TITLE
SQL Dialects

### ISSUE THREAD

**Issue by LarnuUK** [PrismJS/prism#2137]
**Language**
SQL is an option in Prism, but each dialect/flavour (i.e. MySQL, PL\SQL (Oracle), Transact-SQL (SQL Server)), etc have their own nuances, functions, keywords. Could individual languages be added for these dialects. These would likely be an `extend` of the sql language.

I'm happy to give T-SQL a go, but don't know if others would be interested in doing so, or if there is interest.

**Additional resources**
[SQL Server technical documentation](https://docs.microsoft.com/en-us/sql/sql-server/?view=sql-server-ver15)
[MySQL Documentation](https://dev.mysql.com/doc/)
[Database PL/SQL Language Reference](https://docs.oracle.com/database/121/LNPLS/toc.htm)
[PostgreSQL](https://www.postgresql.org/docs/manuals/)

**Comment by RunDevelopment** [comment]
To my knowledge, the MySQL Doc page actually has its own modification of the base Prism SQL language, so one might be able to get most patterns from them.

And we already have a PL/SQL language, but it might not be perfect/up to date.

**Comment by LarnuUK** [comment]
> And we already have a PL/SQL language, but it might not be perfect/up to date.


Hadn't noticed the PL/SQL was listed. I'll have a look at doing a T-SQL one then.

**Comment by IssacEinstein** [comment]
I see a lot of Keywords Missing in the MSSQL. Will post them.
Meanwhile, I had raised an issue [https://github.com/PrismJS/prism/issues/3007]

**Comment by LarnuUK** [comment]
It likely makes sense, if I am honest, while having a "generic" SQL that respects the syntax that is more common to all dialects.  Highlight.JS a separate one for PL/SQL and pgsql, and I *think* they were considering doing one for T-SQL.

I would, however, suggest that for (Microsoft) SQL Server and Oracle's SQL dialect you use their actual names: which are Transact-SQL (T-SQL) and Procedure Language SQL (PL/SQL) respectively. So would be `Prism.languages.plsql` for example.

**Comment by RunDevelopment** [comment]
> Are y'all open to having separate SQL dialects be different languages with their own config object

Yes, it has to be that way. There is no guarantee that these different dialects are even compatible with each other. Sure, they are all based off vanilla SQL, but the same (or very similar) syntax might be used to denote different things in different dialects.

This also works better for users. If a user knows that they're using a specific dialect, we should probably respect that and give them the best highlighting we can instead of say: "You'll get SQL. Take it, or leave it."

Also, it's just easier to implement them as separate languages. Having x-many languages mushed into one grammar is hellish to maintain. Languages evolve and change over time, so we'll have to adjust them. 

> I'm wondering if y'all would merge it.

Of course! Language contributions are highly appreciated. We'll also review your language before merging, so we might even be able to improve it a little :)

**Comment by karlhorky** [comment]
Would also love to have SQL dialects as part of Prism!

For example PostgreSQL `GENERATED` and `ALWAYS` keywords:

Prism:

![Screenshot 2023-11-28 at 09 49 17](https://github.com/PrismJS/prism/assets/1935696/34480a9b-2d66-47a4-a633-a432c29991e4)

VS Code:

![Screenshot 2023-11-28 at 09 49 03](https://github.com/PrismJS/prism/assets/1935696/6aeff9fe-9a16-4b39-8c47-6ace84a937e6)
